rapids-cmake/cpm/gtest.cmake

```diff
@@ -0,0 +1,81 @@
+#=============================================================================
+# Copyright (c) 2020-2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#=============================================================================
+include_guard(GLOBAL)
+
+#[=======================================================================[.rst:
+rapids_cpm_gtest
+----------------
+
+.. versionadded:: v21.10.00
+
+Allow projects to find or build `Google Test` via `CPM` with built-in
+tracking of these dependencies for correct export support.
+
+Uses the version of GTest :ref:`specified in the version file <cpm_versions>` for consistency
+across all RAPIDS projects.
+
+.. code-block:: cmake
+
+  rapids_cpm_gtest( [BUILD_EXPORT_SET <export-name>]
+                    [INSTALL_EXPORT_SET <export-name>]
+                  )
+.. note::
+  Installation of GTest will occur if an INSTALL_EXPORT_SET is provided, and GTest
+  is added to the project via :cmake:command:`add_subdirectory` by CPM.
+
+Result Targets
+^^^^^^^^^^^^^^
+  GTest::gtest, GTest::gmock, GTest::gtest_main, GTest::gmock_main targets will be created
+
+Result Variables
+^^^^^^^^^^^^^^^^
+  :cmake:variable:`GTest_SOURCE_DIR` is set to the path to the source directory of GTest.
+  :cmake:variable:`GTest_BINAR_DIR`  is set to the path to the build directory of  GTest.
+  :cmake:variable:`GTest_ADDED`      is set to a true value if GTest has not been added before.
+
+#]=======================================================================]
+function(rapids_cpm_gtest)
+  list(APPEND CMAKE_MESSAGE_CONTEXT "rapids.cpm.gtest")
+
+  set(to_install OFF)
+  if(INSTALL_EXPORT_SET IN_LIST ARGN)
+    set(to_install ON)
+  endif()
+
+  include("${rapids-cmake-dir}/cpm/detail/package_details.cmake")
+  rapids_cpm_package_details(GTest version repository tag)
+
+  include("${rapids-cmake-dir}/cpm/find.cmake")
+  rapids_cpm_find(GTest ${version} ${ARGN}
+                  GLOBAL_TARGETS GTest::gtest GTest::gmock GTest::gtest_main GTest::gmock_main
+                  CPM_ARGS
+                  GIT_REPOSITORY ${repository}
+                  GIT_TAG ${tag}
+                  GIT_SHALLOW TRUE
```

Comment:
won't `GIT_SHALLOW` shallow being `TRUE` interfere with the ability of setting any commit hash in the json file? Setting it to false would give us more flexibility, no?

Reply:
Since it is always combined with a tag it will do a shallow checkout at the tip of that tag/branch.

Reply:
Ohh I see what you mean, any arbitrary `hash` value for the GIT_TAG.

Yes in those case we need a shallow equals false. I think the best approach is to extend the json spec to include 
`git_shallow` boolean field with the default being `true`

Reply:
I have updated the PR with the optional `git_shallow` field